core/filelocation.cpp

```diff
@@ -0,0 +1,235 @@
+	QString repository = fn.left(branch_from);				// Guaranteed OK owing to isGitFilename() == true
+	QString branch = fn.mid(branch_from + 1, branch_to - branch_from - 1);	// ditto
+	return std::make_pair(repository, branch);
+}
+
+FileLocation::FileLocation(Type type_, const QString &name_, const QString &branch_) : type(type_), name(name_), branch(branch_)
+{
+	if (type == NONE)
+		return;
+	fixName();
+}
+
+FileLocation::FileLocation(Type type_, const QString &name_) : type(type_)
+{
+	if (type == NONE)
+		return;
+	if (type == LOCAL_FILE) {
+		name = name_;
+	} else {
+		// Is a git repository
+		std::pair<QString, QString> repo_branch = parseGitFilename(name_);
+		name = repo_branch.first;
+		branch = repo_branch.second;
+	}
+	fixName();
+}
+
+FileLocation::FileLocation(const QString &type_, const QString &name_, const QString &branch_) : name(name_), branch(branch_)
+{
+	if (type_ == "none")
+		type = NONE;
+	else if (type_ == "local")
+		type = LOCAL_FILE;
+	else if (type_ == "cloud")
+		type = CLOUD_GIT;
+	else if (type_ == "cloud_offline")
+		type = CLOUD_GIT_OFFLINE;
+	else if (type_ == "git")
+		type = GIT;
+	else
+		*this = guessFromFileName(name_); // Unknown or empty type -> try to guess from filename
+	fixName();
+}
+
+FileLocation FileLocation::guessFromFileName(const QString &name)
+{
+	if (name.simplified().isEmpty()) {
+		return FileLocation();
+	} else if (isGitFilename(name)) {
+		std::pair<QString, QString> repo_branch = parseGitFilename(name);
+		// TODO: Check for git repository if local!!
+		return FileLocation(GIT, repo_branch.first, repo_branch.second);
+	}
+	return FileLocation(LOCAL_FILE, name, QString());
+}
+
+QString FileLocation::getName() const
+{
+	return name;
+}
+
+QString FileLocation::getBranch() const
+{
+	return branch;
+}
+
+QString FileLocation::getUser() const
+{
+	return user;
+}
+
+FileLocation::Type FileLocation::getType() const
+{
+	return type;
+}
+
+bool FileLocation::isNone() const
+{
+	return type == NONE;
+}
+
+bool FileLocation::isRemote() const
+{
+	if (type == CLOUD_GIT)
+		return true;
+	if (type != GIT)
+		return false;
+	return isRemoteRepository(name);
+}
+
+bool FileLocation::isCloud() const
+{
+	return type == CLOUD_GIT || type == CLOUD_GIT_OFFLINE;
+}
+
+QString FileLocation::typeAsString() const
+{
+	switch(type) {
+	default:
+	case NONE:		return QString("none");
+	case LOCAL_FILE:	return QString("local");
+	case CLOUD_GIT:		return QString("cloud");
+	case CLOUD_GIT_OFFLINE:	return QString("cloud_offline");
+	case GIT:		return QString("git");
+	}
+}
+
+QString FileLocation::formatShort() const
+{
+	switch(type) {
+	default:
+	case NONE:		return QFileInfo(prefs.default_filename).fileName();
+	case LOCAL_FILE:	return QFileInfo(name).fileName();
+	case CLOUD_GIT:		return QObject::tr("[cloud storage for] %1").arg(branch);
+	case CLOUD_GIT_OFFLINE:	return QObject::tr("[local cache for] %1").arg(branch);
+	case GIT:		return QObject::tr("[%1] in repository %2").arg(branch, repoName());
```

Comment:
ok, I can see why you do all five now. I'm still not convinced that this is adding value, but at least I understand what you are doing

Reply:
As I said above: this is *not* adding value, this is making explicit current practice.

Reply:
😇